### ISSUE TITLE
Nullable reference types

### ISSUE THREAD

**Issue by hankovich** [vkhorikov/CSharpFunctionalExtensions#341]
Recently a lot of code in `Maybe<T>` was changed to use `GetValueOrDefault`, but that method is not annotated properly in terms of NRT attributes. We could add a [[MaybeNull](docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.maybenullattribute)] on top of it.
Also we can do the same for parameters of `Deconstruct` methods for both `Result`'s and `Maybe`. Propapbly there are more places where those attributes are applicable.

NRTs can also be used in projects targeting `netstandard2.0` and legacy .NET Frameworks (for example, [this article explains](https://www.meziantou.net/how-to-use-nullable-reference-types-in-dotnet-standard-2-0-and-dotnet-.htm) how).

**Comment by vkhorikov** [comment]
Yes, we should do this, and it's quite a big chunk of work: we need to enable nullable reference types for the library (to take advantage of the new C# 8 nullable reference types analysis) and carefully go through both `Maybe` and `Result` and put `[MaybeNull]` to all properties that might be nulls there.

Related issues: 
https://github.com/vkhorikov/CSharpFunctionalExtensions/issues/253 https://github.com/vkhorikov/CSharpFunctionalExtensions/issues/139

**Comment by A9G-Data-Droid** [comment]
Isn't Nullable and Maybe addressing the same concern? Why would I choose Maybe over traditional nullability?

**Comment by vkhorikov** [comment]
`Nullable` is for structs only, whereas `Maybe` can be used for both.

And for reference, `Maybe` now supports proper C# nullability annotations.
